File: csharp/strain/Strain.cs
using System;
using System.Collections.Generic;

public static class Strain
{
    public static IEnumerable<T> Keep<T>(this IEnumerable<T> collection, Func<T, bool> predicate)
    {
        foreach (var item in collection)
        {
            if (predicate.Invoke(item))
            {
                yield return item;
            }
        }
    }

    public static IEnumerable<T> Discard<T>(this IEnumerable<T> collection, Func<T, bool> predicate)
    {
        foreach (var item in collection)
        {
            if (!predicate.Invoke(item))
            {
                yield return item;
            }
        }
    }
}

File: scala/strain/src/main/scala/Strain.scala
object Strain {
  def keep[T](items: Seq[T], filter: T => Boolean): Seq[T] = {
    for (elem <- items if filter(elem)) yield elem
  }

  def discard[T](items: Seq[T], filter: T => Boolean): Seq[T] = {
    for (elem <- items if !filter(elem)) yield elem
  }
}